board support: ZC702 LS1021ATWR

yanyan-wrs opened the pull request:
-Xilinx Zynq 7000 ZC702 (plat-zynq)
Update board support
  -NXP LS1021ATWR (plat-ls)
Signed-off-by: Yan Yan yan.yan@windriver.com

etienne-lms commented on core/arch/arm/include/sm/optee_smc.h:
```diff
@@ -385,6 +385,21 @@
+ */
+#define OPTEE_SMC_FUNCID_PLATFORM_DEFINED  0x100
+#define OPTEE_SMC_PLATFORM_DEFINED \
+	OPTEE_SMC_FAST_CALL_VAL(OPTEE_SMC_FUNCID_PLATFORM_DEFINED)
```
This is a bit ugly. Should find a clean way to define platform specific smc's.

yanyan-wrs replied:
You're correct, those are the right categories to use.

etienne-lms replied:
As it is platform specific, could it be defined in a `plat-xxx/` file?

etienne-lms commented on core/arch/arm/kernel/generic_boot.c:
```diff
@@ -74,6 +74,7 @@
 
 #ifdef CFG_BOOT_SECONDARY_REQUEST
 uint32_t ns_entry_addrs[CFG_TEE_CORE_NB_CORE] __data;
+uint32_t spin_table[CFG_TEE_CORE_NB_CORE] __data = {0};
```
Don't need the `= { 0 };`.

yanyan-wrs replied:
fixed

etienne-lms commented on core/arch/arm/plat-imx/a9_plat_init.S:
```diff
@@ -188,6 +188,22 @@ END_FUNC arm_cl2_enable
 
+	read_mpidr r0
+	ands r0, #3
+	bne 1f
```
Why should secondary not init scr?

yanyan-wrs replied:
This is indeed a BUG, all cores shall program its SCR.

etienne-lms commented on core/arch/arm/plat-imx/main.c:
```diff
@@ -82,6 +82,92 @@ static void main_fiq(void)
+{
+	uintptr_t addr;
+
+	if (0 == (read_mpidr() & 0x3)) {
```
Use `get_core_pos().`

etienne-lms commented on core/arch/arm/plat-ls/main.c:
```diff
@@ -94,3 +167,36 @@ void console_flush(void)
+	itr_init(&gic_data.chip);
+}
+
+static void platform_tee_entry_fast(struct thread_smc_args *args)
```
This is generic. Could'it be implemented in a generic source file?

yanyan-wrs replied:
I am not sure, this function happens to be identical for both i.MX6 and LS102x; but on ZC702, it extends this function to handle additional platform defined SMCs. So this function is intended to be extended or overwritten for each platform, it may call other "generic" functions (e.g., `the generic_smp_boot`) if that platform happens to support a common SMC request, but the function overall is platform dependent thus not generic.

etienne-lms replied:
`OPTEE_SMC_BOOT_SECONDARY` is a generic OPTEE SMC.
Could you change `tee_entry_fast()` to handle `OPTEE_SMC_BOOT_SECONDARY` and call `generic_smp_boot()` when `CFG_BOOT_SECONDARY_REQUEST` is enable ?
This will factorize platform code (`plat-imx/main.c` to update).

yanyan-wrs replied:
I see your point.

etienne-lms commented on core/arch/arm/plat-zynq/main.c:
```diff
@@ -0,0 +1,290 @@
+
+vaddr_t pl310_base(void)
+{
+	static void *va __data; /* in case it's used before .bss is cleared */
```
Use `__early_bss` instead of `__data.`

yanyan-wrs replied:
fixed

etienne-lms commented on core/arch/arm/plat-zynq/main.c:
```diff
@@ -0,0 +1,290 @@
+	return PL310_BASE;
+}
+
+void main_init_gic(void)
```
Take care. Secondary core shall call `gic_cpu_init().`

yanyan-wrs replied:
GIC init for each core is actually done here in `platform_cpu_reset_late()` in the same file. Anyway, I'll use the driver interface for consistency.
Is it added recently? the last time I check there is no `main_secondary_init_gic` hook in generic boot code.

etienne-lms replied:
yes. the `gic_cpu_init()` existed but was not used. this function also load gic_ctrl register.
and as you said, better for consistency and maintenance.

etienne-lms commented on core/arch/arm/plat-zynq/main.c:
```diff
@@ -0,0 +1,290 @@
+
+	for (i = 0; i < ARRAY_SIZE(slcr_access_range); i+=2) {
+		if (addr >= slcr_access_range[i] &&
+		addr <= slcr_access_range[i+1]) {
```
Minor: identation

yanyan-wrs replied:
should I align the second 'addr' at the same column of the first 'addr', and it means adding a few spaces instead of tab? Sorry I am not familiar with this.

etienne-lms replied:
```
if (xxx &&
    yyy) {             <---- here spaces to align with the parenthesis (eventually a tabulation)
        instructions...       <---- here 1 tab indentation.
}
```

etienne-lms commented on core/arch/arm/plat-zynq/main.c:
```diff
@@ -0,0 +1,290 @@
+		if (addr >= slcr_access_range[i] &&
+		addr <= slcr_access_range[i+1]) {
+			static vaddr_t va __data;
+			if (!va)
```
Minor: add an empty line between variables and instructions.

yanyan-wrs replied:
fixed

etienne-lms commented on core/arch/arm/plat-zynq/plat_init.S:
```diff
@@ -0,0 +1,253 @@
+
+	read_mpidr r0
+	ands r0, #3
+	bne 1f
```
Secondary boot cores do not init SCR?

yanyan-wrs replied:
BUG; fixed.

etienne-lms commented on core/arch/arm/plat-zynq/main.c:
```diff
@@ -0,0 +1,290 @@
+
+void plat_cpu_reset_late(void)
+{
+	if (0 == (read_mpidr() & 0x3)) {
```
Prefer `if (get_core_pos()) { ... }`

etienne-lms commented on core/arch/arm/plat-zynq/main.c:
```diff
@@ -0,0 +1,290 @@
+#if defined (CFG_BOOT_SECONDARY_REQUEST)
+		/* set secondary entry address and release core */
+		write32(CFG_TEE_LOAD_ADDR, 0xFFFFFFF0);
+		asm ("dsb");
```
Use the asm32.h routines `dmb(),` `sev().`. Add them if they do not exist.

etienne-lms commented on core/arch/arm/kernel/generic_boot.c:
```diff
@@ -662,3 +663,20 @@ void generic_boot_init_secondary(unsigned long nsec_entry)
+
+	/* set secondary cores' NS entry addresses */
+
+	ns_entry_addrs[core_idx] = entry;
```
You should vaditate the entry point address.
Verify at least that it is not inside TEE RAM or TA RAM.

yanyan-wrs replied:
Actually this entry point address is used to fill the NS context which the CPU is about to switch to. If the provided address is in TEE RAM or TA RAM, it causes exception the same as the NS execution tries to reference Secure memory. So at least no secure leak is exposed, thus no need to check it against TEE RAM or TA RAM.

etienne-lms replied:
you are right. no need for extra checks.

etienne-lms commented on core/arch/arm/plat-imx/main.c:
```diff
@@ -82,6 +82,92 @@ static void main_fiq(void)
+		/* primary core */
+#if defined (CFG_BOOT_SECONDARY_REQUEST)
+		/* set secondary entry address and release core */
+		write32(CFG_TEE_LOAD_ADDR, SRC_BASE+SRC_GPR1 + 8);
```
Minor: space around '+'

yanyan-wrs replied:
fixed

etienne-lms replied:
also a space to remove line 93:
```
-#if defined (CFG_BOOT_SECONDARY_REQUEST)
+#if defined(CFG_BOOT_SECONDARY_REQUEST)
```

yanyan-wrs replied:
fixed, also caught by checker script.

jbech-linaro commented on .travis.yml:
```diff
@@ -201,6 +201,9 @@ script:
   - $make PLATFORM=ls-ls1021atwr
   - $make PLATFORM=ls-ls1021aqds
 
+  # Xilinx Zynq 7000
```
Without knowing too much about Xilinx platforms, is "7000" and "zc702" considered as two different names for the same hardware?

yanyan-wrs replied:
Zynq-7000 is the processor family, and zc702 is the board name; there are also zc706 and ZED board:
http://china.xilinx.com/products/boards-and-kits/ek-z7-zc702-g.html

jenswi-linaro commented:
This change is touching on generic code and three platforms. Is there a way you could split it in a few separate commits (OK to have them in the same pull request)?

yanyan-wrs commented on core/arch/arm/plat-ls/main.c:
```diff
@@ -61,13 +72,75 @@ static void main_fiq(void)
+			 addr += 4)
+			write32(read32(addr) | __compiler_bswap32(0x01000100), addr);
+	} else {
+	/* program the cntfrq, the cntfrq is banked for each core */
```
fixed

yanyan-wrs commented on core/arch/arm/plat-ls/main.c:
```diff
@@ -61,13 +72,75 @@ static void main_fiq(void)
+
+#if defined (CFG_BOOT_SECONDARY_REQUEST)
+		/* set secondary entry address */
+		write32(__compiler_bswap32(CFG_TEE_LOAD_ADDR),
```
I'll wait for this patch.

yanyan-wrs commented:
@jenswi-linaro I'll fix the issues already pointed out, and then split this into several commits.

etienne-lms commented on core/arch/arm/plat-zynq/main.c:
```diff
@@ -0,0 +1,290 @@
+static uint32_t read_slcr(uint32_t addr, uint32_t * val)
+{
+	uint32_t i;
+	for (i = 0; i < ARRAY_SIZE(slcr_access_range); i+=2) {
```
add 1 empty line

etienne-lms commented on core/arch/arm/plat-zynq/main.c:
```diff
@@ -0,0 +1,290 @@
+	uint32_t i;
+	for (i = 0; i < ARRAY_SIZE(slcr_access_range); i+=2) {
+		if (addr >= slcr_access_range[i] &&
+		addr <= slcr_access_range[i+1]) {
```
same indentation

etienne-lms commented on core/arch/arm/plat-zynq/main.c:
```diff
@@ -0,0 +1,290 @@
+		addr <= slcr_access_range[i+1]) {
+			static vaddr_t va __data;
+			if (!va)
+				va = (vaddr_t)phys_to_virt(SLCR_BASE,
```
1 empty line + `__early_bss` instead of `__data`

etienne-lms commented on core/arch/arm/plat-zynq/plat_init.S:
```diff
@@ -0,0 +1,253 @@
+	 * to 0x00020202 for proper cache operations.
+	 */
+	ldr  r2, =ZYNQ_SLCR_L2C_RAM
+	ldr  r1, =0x00020202
```
magic init values will better fit in the `platform_config.h`

yanyan-wrs commented on core/arch/arm/kernel/generic_entry_a32.S:
```diff
@@ -473,6 +473,16 @@ UNWIND(	.cantunwind)
+	ldr r0, [r1]
+	cmp	r0, #0
+	beq	spin
+
```
not really, what if the NS entry is indeed at address 0x0 (who knows what hardware can come up with), so we cannot check `ns_entry_addrs` against 0 for that matter.

etienne-lms replied:
i was thinking to something like:
```
	/* spin until secondary boot request */
	bl      get_core_pos
	lsl     r0, #2                /* size of uint32_t */
	ldr     r1, =spin_table

spin:
	wfe
	ldr     r2, [r1, r0]
	cmp     r2, #0
	beq     spin

	/* get ns entry from ns_entry_addrs array */
	dmb
 	ldr     r1, =ns_entry_addrs
	ldr     r0, [r1, r0]
```

Take care to the `dmb.` I think it is required to not use a value loaded from a predictive execution.

yanyan-wrs replied:
You're correct, the `dmb` is a must to sync the value set and get between cores. How could I have missed this!

etienne-lms commented on core/arch/arm/kernel/generic_boot.c:
```diff
@@ -662,3 +663,20 @@ void generic_boot_init_secondary(unsigned long nsec_entry)
+	/* set secondary cores' NS entry addresses */
+
+	ns_entry_addrs[core_idx] = entry;
+	spin_table[core_idx] = 1;
```
i think you need a `dmb` so that spin is updated before ns entry address reach a cache.

yanyan-wrs commented:
I changed the code according to your review comments, and split the PR into several commits:
1. add the generic secondary boot in common code, and also the i.MX6 which is the only existing board using the secondary boot interface.
2. modify the ls1021atwr board to support dual core, and other minor changes.

The support for Xilinx ZC702 board will be in another PR after this one is merged.

etienne-lms commented on core/arch/arm/tee/entry_fast.c:
```diff
@@ -164,6 +175,11 @@ void tee_entry_fast(struct thread_smc_args *args)
 	case OPTEE_SMC_ENABLE_SHM_CACHE:
 		tee_entry_enable_shm_cache(args);
 		break;
+#ifdef CFG_BOOT_SECONDARY_REQUEST
```
remove the `#ifdef` here. function`tee_entry_boot_secondary()` shall be define in both cases.
see comment above.

yanyan-wrs replied:
fixed

etienne-lms commented on core/arch/arm/tee/entry_fast.c:
```diff
@@ -127,6 +128,16 @@ static void tee_entry_enable_shm_cache(struct thread_smc_args *args)
+		args->a0 = OPTEE_SMC_RETURN_OK;
+	else
+		args->a0 = OPTEE_SMC_RETURN_EBADCMD;
+#endif
```
if `CFG_BOOT_SECONDARY_REQUEST` is not enabled, `tee_entry_boot_secondary()` shall load `OPTEE_SMC_RETURN_UNKNOWN_FUNCTION` or not-implemented, if this exists.

etienne-lms commented on core/arch/arm/plat-ls/main.c:
```diff
@@ -94,3 +152,27 @@ void console_flush(void)
+
+	/* Initialize GIC */
+	gic_init(&gic_data, gicc_base, gicd_base);
+
```
minor: remove empty line

yanyan-wrs replied:
fixed

etienne-lms commented on core/arch/arm/plat-zynq7k/main.c:
```diff
@@ -0,0 +1,270 @@
+#include <platform_config.h>
+#include <platform_smc.h>
+#include <stdint.h>
+#include <tee/entry_std.h>
```
minor: swap these: `fast` before `std`

yanyan-wrs replied:
fixed

etienne-lms commented on core/arch/arm/plat-zynq7k/main.c:
```diff
@@ -0,0 +1,270 @@
+
+void plat_cpu_reset_late(void)
+{
+	if (get_core_pos() == 0) {
```
prefer `if (!get_core_pos())`.

yanyan-wrs replied:
fixed

etienne-lms commented on core/arch/arm/plat-zynq7k/main.c:
```diff
@@ -0,0 +1,270 @@
+		sev();
+#endif
+
+		/*
```
could remove this description

yanyan-wrs replied:
fixed

etienne-lms commented on core/arch/arm/plat-zynq7k/main.c:
```diff
@@ -0,0 +1,270 @@
+			SCU_BASE + SCU_CTRL);
+
+		/* NS Access control */
+
```
spurious empty line

etienne-lms commented on core/arch/arm/plat-zynq7k/main.c:
```diff
@@ -0,0 +1,270 @@
+		write32(0x00000FFF, SCU_BASE + SCU_NSAC);
+
+		/* SCU enable */
+
```
empty line

etienne-lms commented on core/arch/arm/plat-zynq7k/main.c:
```diff
@@ -0,0 +1,270 @@
+
+	/* Initialize GIC */
+	gic_init(&gic_data, gicc_base, gicd_base);
+
```
empty line

etienne-lms commented on core/arch/arm/plat-zynq7k/plat_init.S:
```diff
@@ -0,0 +1,248 @@
+ * Trap CPU in case of error.
+ */
+FUNC arm_cl2_config , :
+UNWIND(	.fnstart)
```
spurious tabulation

yanyan-wrs replied:
fixed

etienne-lms commented on core/arch/arm/plat-zynq7k/plat_init.S:
```diff
@@ -0,0 +1,248 @@
+	movw r1, #0xF999
+	movt r1, #0xFFFF
+	orr  r2,r2,r1
+	str  r2, [r0, #PL310_TAG_RAM_CTRL]
```
check latest changes in `plat-stm/`. could help you clean this a bit if you want.
jens pointed out some useless operations in this legacy sequence.

yanyan-wrs replied:
I think I'll keep the L2 cache config code here, in case the bootloader (u-boot) does not set it properly.

etienne-lms commented on core/arch/arm/plat-zynq7k/platform_smc.h:
```diff
@@ -0,0 +1,81 @@
+ * a0	OPTEE_SMC_RETURN_EBADCMD
+ * a1	Undefined value
+ * a2-3	Not used
+ * a4-7	Preserved
```
comments are clear but a0/a1/... use different indentations.

here and in the other cmd below.

yanyan-wrs replied:
fixed

etienne-lms commented on core/arch/arm/tee/entry_fast.c:
```diff
@@ -128,15 +128,16 @@ static void tee_entry_enable_shm_cache(struct thread_smc_args *args)
 		args->a0 = OPTEE_SMC_RETURN_EBADCMD;
-#endif
 }
+#endif
```
no, prefer always define the function.

yanyan-wrs replied:
fixed

yanyan-wrs commented on core/arch/arm/include/kernel/generic_boot.h:
```diff
@@ -56,6 +56,8 @@ void arm_cl2_enable(vaddr_t pl310);
 #ifdef CFG_BOOT_SECONDARY_REQUEST
 extern uint32_t ns_entry_addrs[];
+extern uint32_t spin_table[];
+int generic_boot_core(size_t core_idx, uint32_t entry);
```
fixed

yanyan-wrs commented on core/arch/arm/kernel/generic_boot.c:
```diff
@@ -662,3 +663,21 @@ void generic_boot_init_secondary(unsigned long nsec_entry)
+	ns_entry_addrs[core_idx] = entry;
+	spin_table[core_idx] = 1;
+
+	dsb();
```
fixed

yanyan-wrs commented on core/arch/arm/kernel/generic_entry_a32.S:
```diff
@@ -473,10 +473,22 @@ UNWIND(	.cantunwind)
 	bl	cpu_is_ready
 
 #if defined (CFG_BOOT_SECONDARY_REQUEST)
+spin:
```
fixed

yanyan-wrs commented on core/arch/arm/plat-imx/main.c:
```diff
@@ -82,6 +82,77 @@ static void main_fiq(void)
+{
+	uintptr_t addr;
+
+	if (get_core_pos() == 0) {
```
fixed

etienne-lms commented on core/arch/arm/include/kernel/generic_boot.h:
```diff
@@ -54,9 +54,8 @@ void plat_cpu_reset_late(void);
-extern uint32_t ns_entry_addrs[];
-#endif
+int generic_boot_core_release(size_t core_idx, paddr_t entry);
+paddr_t generic_boot_core_hpen(void);
```
if `CFG_BOOT_SECONDARY_REQUEST` is not enable, `generic_boot_core_xxx()` should not be defined. otherwise they should panic but it is not clean to embed code that only aims at panicking.

yanyan-wrs replied:
fixed, defines protected using CFG_BOOT_SECONDARY_REQUEST.

etienne-lms commented on core/arch/arm/kernel/generic_boot.c:
```diff
@@ -662,3 +661,30 @@ void generic_boot_init_secondary(unsigned long nsec_entry)
 }
 #endif
+
+int generic_boot_core_release(size_t core_idx, paddr_t entry)
```
`#if defined(CFG_BOOT_SECONDARY_REQUEST)`

yanyan-wrs replied:
As you pointed out, this function shall be protected by CFG_BOOT_SECONDARY_REQUEST, it is not used in build, so no need to do panic on the fly.

yanyan-wrs replied:
fixed

etienne-lms commented on core/arch/arm/tee/entry_fast.c:
```diff
@@ -127,6 +128,14 @@ static void tee_entry_enable_shm_cache(struct thread_smc_args *args)
 }
 
+static void tee_entry_boot_secondary(struct thread_smc_args *args)
+{
```
```
#if defined(CFG_BOOT_SECONDARY_REQUEST)
	if (!generic_boot_core_release(...))
		args->a0 = OPTEE_SMC_RETURN_OK;
	else
		args->a0 = OPTEE_SMC_RETURN_EBADCMD;
#else
	args->a0 = OPTEE_SMC_RETURN_ENOTAVAIL;
#endif
```

yanyan-wrs replied:
I am changing it to this.

etienne-lms commented on core/arch/arm/kernel/generic_boot.c:
```diff
@@ -72,9 +72,8 @@
-#ifdef CFG_BOOT_SECONDARY_REQUEST
-uint32_t ns_entry_addrs[CFG_TEE_CORE_NB_CORE] __data;
-#endif
+static paddr_t ns_entry_addrs[CFG_TEE_CORE_NB_CORE] __early_bss;
```
preserve `#if defined(CFG_BOOT_SECONDARY_REQUEST)`.
Otherwise these variables are uselessly linked in the early bss section.

yanyan-wrs replied:
yes, I shall protect all of them.

etienne-lms commented on core/arch/arm/plat-ls/plat_init.S:
```diff
@@ -0,0 +1,99 @@
+.section .data
+.balign 4
+
+cntfrq:
```
is this `cntfrq` used ?

yanyan-wrs replied:
deleted.

yanyan-wrs commented on core/arch/arm/kernel/generic_entry_a32.S:
```diff
@@ -473,11 +473,7 @@ UNWIND(	.cantunwind)
-	ldr r1, =ns_entry_addrs
-	add r1, r1, r0, lsl #2	/* size of uint32_t */
-	ldr r0, [r1]
+	bl	generic_boot_core_hpen
```
fixed

yanyan-wrs commented on core/arch/arm/plat-imx/main.c:
```diff
@@ -82,6 +82,77 @@ static void main_fiq(void)
+		write32(CFG_TEE_LOAD_ADDR, SRC_BASE+SRC_GPR1+16);
+		write32(CFG_TEE_LOAD_ADDR, SRC_BASE+SRC_GPR1+24);
+
+		write32(0x7 << SRC_SCR_ENABLE_OFFSET, SRC_BASE+SRC_SCR);
```
write32(CFG_TEE_LOAD_ADDR, SRC_BASE + SRC_GPR1 + 16);
Is above right? I am not sure if the space between '+' and constant is needed?

yanyan-wrs commented on core/arch/arm/plat-ls/main.c:
```diff
@@ -61,13 +72,60 @@ static void main_fiq(void)
 
+void plat_cpu_reset_late(void)
+{
+	static uint32_t cntfrq;
```
Thanks for the careful review.
correct, it should be in __early_bss;
however, the secondary cores use it before core0 enters main(), so it actually works, although it shall be placed in __early_bss.

yanyan-wrs commented on core/arch/arm/plat-ls/main.c:
```diff
@@ -61,13 +72,60 @@ static void main_fiq(void)
+
+	if (!get_core_pos()) {
+		/* read cnt freq */
+
```
fixed

yanyan-wrs commented on core/arch/arm/plat-zynq7k/main.c:
```diff
@@ -0,0 +1,273 @@
+		dsb();
+		sev();
+#endif
+		/* Invalidate all registers */
```
fixed

etienne-lms commented:
@yanyan-wrs thanks for the changes. content is ok.
Could you push back you changes as 3 separated commits, rebased on master branch?

etienne-lms commented:
Could it be possible to split the first commit in 2 ?
it mixes generic 'secondary_boot' and 'plat-imx/' inits re-factoring.

when possible, it's better to have specific commits for specific features or changes, especially when several files are modified.

yanyan-wrs commented:
Sure, I'll split the commit as you requested.

yanyan-wrs commented on core/arch/arm/kernel/generic_boot.c:
```diff
@@ -662,3 +663,32 @@ void generic_boot_init_secondary(unsigned long nsec_entry)
+#if defined(CFG_BOOT_SECONDARY_REQUEST)
+int generic_boot_core_release(size_t core_idx, paddr_t entry)
+{
+	if ((!core_idx) || (core_idx >= CFG_TEE_CORE_NB_CORE))
```
fixed

yanyan-wrs commented:
Could someone please merge this, or is there something missing?

jforissier commented:
@yanyan-wrs please fix the conflicts with master (i.e., rebase). Then I'll need some Reviewed-by tags.

yanyan-wrs commented on core/arch/arm/plat-imx/platform_config.h:
```diff
@@ -113,54 +113,141 @@
+#define SRC_BASE			0x020D8000
+#define SRC_SCR				0x000
+#define SRC_GPR1			0x020
+#define SRC_SCR_CPU_ENABLE_ALL		(0x7 << 22)
```
fixed

yanyan-wrs commented on core/arch/arm/plat-imx/main.c:
```diff
@@ -177,6 +176,22 @@ vaddr_t pl310_base(void)
 	return PL310_BASE;
 }
 
+void arm_cl2_config(vaddr_t pl310 __unused)
```
this `arm_cl2_config` is only used before mmu enabled, the physical addr is fine.

yanyan-wrs commented on core/arch/arm/plat-imx/a9_plat_init.S:
```diff
@@ -189,6 +86,17 @@ FUNC plat_cpu_reset_early , :
+	 * Secure World [bit0: NS=0]
+	 */
+	mov r0, #SCR_AW
+	write_scr r0		/* write Secure Configuration Register */
```
removed

yanyan-wrs commented:
Conflicts merged.

jenswi-linaro commented:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

jforissier commented on core/arch/arm/plat-zynq7k/plat_init.S:
```diff
@@ -0,0 +1,110 @@
+.code 32
+
+/*
+ * Cortex A9 configuration early configuration
```
Remove first occurrence of 'configuration'

jforissier commented on core/arch/arm/plat-zynq7k/plat_init.S:
```diff
@@ -0,0 +1,110 @@
+/*
+ * Cortex A9 configuration early configuration
+ *
+ * Use scratables registers R0-R3.
```
scratables?

jforissier commented on core/arch/arm/plat-zynq7k/platform_config.h:
```diff
@@ -0,0 +1,274 @@
+ *
+ * Invalidate all registers
+ */
+#define	SCU_INV_CTRL_INIT		0xFFFFFFFF
```
Space instead of tab between #define and identifier, please

jforissier commented on core/arch/arm/plat-zynq7k/platform_config.h:
```diff
@@ -0,0 +1,274 @@
+#define SCU_NSAC_CTRL_INIT		0x00000FFF
+
+/* all bit enabled in access control register */
+#define	ACCESS_BITS_ALL			0xFFFFFFFF
```
Same as above

jforissier commented on core/arch/arm/plat-zynq7k/platform_config.h:
```diff
@@ -0,0 +1,274 @@
+ *  TA_RAM  : all what is left (at least 2MByte !)
+ */
+
+/* emulated SRAM, 256K at start of secure DDR */
```
This diagram does not exactly reflect the definitions below, I think you should split it in two (SRAM and DDR).

yanyan-wrs commented:
Fixed typos and diagrams, added to supported board list and maintainers.

jforissier commented:
There are still some inconsistencies in the memory layout diagrams, but that's not a big deal.
For commit "core: arm: support Xilinx ZYNQ7000 ZC702 (plat-zynq7k)":
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier commented:
It looks like this series is basically ready for merge (all comments were addressed I think). To all reviewers who have commented, could you please give your Reviewed-by if you're OK with this?

etienne-lms commented:
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`

jforissier commented:
@yanyan-wrs can you please rebase and add the R-b tags? Then I can merge this. Thanks!

jforissier commented:
@yanyan-wrs I was about to merge this PR when I realized another rebase was needed due to other PRs have been merged recently.
After rebasing I found out that PLATFORM=zynq7k-zc702 would fail to build:
```
  AS      out/arm-plat-zynq7k/core/arch/arm/plat-zynq7k/plat_init.o
core/arch/arm/plat-zynq7k/plat_init.S: Assembler messages:
core/arch/arm/plat-zynq7k/plat_init.S:106: Error: bad instruction `write_pcr r0'
mk/compile.mk:146: recipe for target 'out/arm-plat-zynq7k/core/arch/arm/plat-zynq7k/plat_init.o' failed
make: *** [out/arm-plat-zynq7k/core/arch/arm/plat-zynq7k/plat_init.o] Error 1
```

I think the fix is the following:

```diff
diff --git a/core/arch/arm/plat-zynq7k/plat_init.S b/core/arch/arm/plat-zynq7k/plat_init.S
index eb0be11..6d99a30 100644
--- a/core/arch/arm/plat-zynq7k/plat_init.S
+++ b/core/arch/arm/plat-zynq7k/plat_init.S
@@ -36,6 +36,7 @@
 
 #include <arm32.h>
 #include <arm32_macros.S>
+#include <arm32_macros_cortex_a9.S>
 #include <asm.S>
 #include <kernel/tz_ssvce_def.h>
 #include <kernel/unwind.h>
```
...but I did not want to commit that without your approval. So:

- Can you please rebase and check the above issue?
- While you're at it, would you please update your Git user name so that the author is `Yan Yan <yan.yan@windriver.com>` instead of `yanyan-wrs <yan.yan@windriver.com>`? I think all you need to do is: `git config --global user.name "Yan Yan"`.

Thank you.

jforissier commented:
@yanyan-wrs thanks.
